writer/template/ring.go

```diff
@@ -0,0 +1,257 @@
+
+	"github.com/mauricelam/genny/generic"
+)
+
+//go:generate genny -ast -pkg=writer -in=$GOFILE -out=../datapoint_ring.gen.go -imp "github.com/signalfx/golib/datapoint" gen "Instance=Datapoint:datapoint.Datapoint"
+//go:generate genny -ast -pkg=writer -in=$GOFILE -out=../span_ring.gen.go -imp "github.com/signalfx/golib/trace" gen "Instance=Span:trace.Span"
+
+type Instance generic.Type
+
+// InstancePreprocessor is used to filter out or otherwise change instances
+// before being sent.  If the return value is false, the instance won't be
+// sent.
+type InstancePreprocessor func(*Instance) bool
+
+// InstanceSender is what sends a slice of instances.  It should block until
+// the instances have been sent, or an error has occurred.
+type InstanceSender func(context.Context, []*Instance) error
+
+// InstanceRingWriter is an abstraction that accepts a bunch of instances,
+// buffers them in a ring buffer and sends them out in batches.
+type InstanceRingWriter struct {
+	inputChan    chan *Instance
+	shutdownFlag chan struct{}
+
+	preprocessFunc InstancePreprocessor
+	sendFunc       InstanceSender
+
+	MaxBuffered  int
+	MaxRequests  int
+	MaxBatchSize int
+
+	requestsActive  int64
+	requestsWaiting int64
+	totalWaiting    int64
+
+	totalReceived           int64
+	totalFilteredOut        int64
+	totalInFlight           int64
+	totalSent               int64
+	totalFailedToSend       int64
+	totalPotentiallyDropped int64
+}
+
+// NewInstanceRingWriter returns an initialized but not running instance of the
+// InstanceRingWriter. You must call Run on the returned instance for it to work.
+// preprocessFunc can be used for filtering or modifying instances before
+// being sent.  If preprocessFunc returns false, the instance will not be
+// sent.  preprocessFunc can be nil, in which case all instances will be sent.
+// sendFunc must be provided as the writer is useless without it.
+func NewInstanceRingWriter(preprocessFunc InstancePreprocessor, sendFunc InstanceSender) *InstanceRingWriter {
+	return &InstanceRingWriter{
+		// Give the input channel a bit of buffer, but in general the writer
+		// should be pulling off of it faster than instances would normally be
+		// generated.
+		inputChan:      make(chan *Instance, 1000),
+		preprocessFunc: preprocessFunc,
+		sendFunc:       sendFunc,
+
+		MaxBuffered:  10000,
+		MaxRequests:  10,
+		MaxBatchSize: 1000,
+	}
+}
+
+// InputChan returns the channel that should be used to send instances to this
+// writer.
+func (w *InstanceRingWriter) InputChan() chan *Instance {
+	return w.inputChan
+}
+
+func (w *InstanceRingWriter) WaitForShutdown() {
+	if w.shutdownFlag == nil {
+		panic("should not wait for writer shutdown when not running")
+	}
+	<-w.shutdownFlag
+}
+
+// Run waits for things to come in on the provided
+// channels and forwards them to SignalFx.  This function blocks until the
+// provided context is finished.
+// nolint: dupl
+func (w *InstanceRingWriter) Run(ctx context.Context) {
```

Comment:
how much laggy will the data become as a result of batching?
because of batch how is the response going to reach the correct caller who send in datapoint?